hoomd/test/test_gpu_array.cc

```diff
@@ -16,6 +16,9 @@
 using namespace hoomd::test;
 #endif
 
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wpessimizing-move"
```

Comment:
The 3D code path for GJK started generating warnings when compiled for 2D shapes:
https://github.com/glotzerlab/hoomd-blue/blob/c87d24f91d8b70a40437dedb4d9710226880dcee/hoomd/md/GJK_SV.h#L461-L468

I don't know why the warnings started with this pull request, but on CI I require that we build warning free.

Reply:
We do use C++17 features now, but as this must compile on the GPU we cannot use `if constexpr`. Or at least I assume this is the case.

Reply:
You can use `if constexpr` in device code.

Reply:
@vyas nvcc gives an error as we pass in the c++14 standard (we still support CUDA 10).

In any case, `num_used` is not a constexpr because it depends on the `W_used` input. I added a `else if (ndim==3)` check in that final clause which seems to work as `ndim` is a template parameter known at compile time. Is this change correct?

Reply:
Ah I thought @b-butler was implying that `if constexpr` was not supported by `nvcc` even with C++17. I did not realize that you had upgraded the requirement for only the host compiler. 

Yes, the change that you included is technically equivalent. Unlike with an `if constexpr` you are not guaranteed that the compiler will optimize that branch out, so it's still possible that you'll get errors, but I guess even with -O1 the dead code optimizer is happy to remove branches that are clearly defined at compile time. I assume that you'll still see the warning in debug builds without optimization, but that probably doesn't matter so much.

If you use the `if (ndim == 3)` condition, it would be helpful to add an extra comment there since it may not be obvious why that condition looks different from everything else in the cascade. Something like this might be appropriate (I'm being very verbose and descriptive, feel free to shorten to whatever level of detail that you think is appropriate):
> The maximum number of points used to define a simplex in N dimensions is N+1. Therefore, `num_used` can only be 4 in three dimensions, so this branch is unreachable in 2D code paths. While a simple `else` clause is sufficient for correctness since one of the preceding three branches will always be taken in 2D, this approach is problematic because even compiling `s3d` invocations into 2D code paths leads to compiler warnings. To suppress these warnings, we instead condition this branch on the dimensionality of the system (which is a compile-time constant) so that the dead code optimizer will compile it out and avoid the warnings.

Reply:
Also just FYI my Github handle is @vyasr. You've tagged somebody else in your last message :)

Reply:
> Also just FYI my Github handle is @vyasr. You've tagged somebody else in your last message :)

Yes, I know what your handle is, but GitHub's mandatory auto-complete does not.

> If you use the `if (ndim == 3)` condition, it would be helpful to add an extra comment there since it may not be obvious why that condition looks different from everything else in the cascade. Something like this might be appropriate (I'm being very verbose and descriptive, feel free to shorten to whatever level of detail that you think is appropriate):

Even if we could use `constexpr`, it would still need to be `ndim==3` because `num_used` is not known at compile time. I will add the comment.